### PR TITLE
[nhcb branch] Add basic unit tests for native histograms with custom buckets converted from classic histograms

### PR DESCRIPTION
**Pull request by zenador**
Addresses https://github.com/prometheus/prometheus/issues/13508

~Currently `histogram_quantile` is failing on the newly converted histograms, but the other basic histogram functions work as expected.~ (fixed now)

~Needed to update the behaviour of `histogram_stddev` and `histogram_stdvar` to handle negative bounds properly (previously it was already bugged even for exponential histograms with negative observations).~ (now in https://github.com/prometheus/prometheus/pull/13824)

### CODE REVIEW

File: promql/test.go [hunk 1]
```diff
@@ -584,67 +733,15 @@ func (t *test) exec(tc testCommand, engine engineQuerier) error {
...
-						vec = append(vec, Sample{Metric: series.Metric, T: point.T, F: point.F})
-						break
-					}
-				}
-				for _, point := range series.Histograms {
-					if point.T == timeMilliseconds(iq.evalTime) {
-						vec = append(vec, Sample{Metric: series.Metric, T: point.T, H: point.H})
-						break
-					}
+			if strings.Contains(iq.expr, "_bucket") {
```
Comment (krajorama) [inline promql/test.go:739]: I'd rather we didn't do this, we should make the copy by hand to be explicit. Unit tests should be very clear, explicit and define expectations.

File: promql/test.go [hunk 2]
```diff
@@ -653,6 +750,71 @@ func (t *test) exec(tc testCommand, engine engineQuerier) error {
 	return nil
 }
 
+func (t *test) runQuery(iq atModifierTestCase, cmd *evalCmd, engine engineQuerier) error {
+	// fmt.Printf("expr: %s, evalTime: %v\n", iq.expr, iq.evalTime)
```
[the remark this answers is not in the document]
Comment (zenador) [reply promql/test.go:754]: Noted and will remove eventually, have marked this commit as temporary in the commit message as it's also doing other temporary things like removing failing tests (to make it clear which are failing). Regardless of how we decide to handle this later, this commit is useful for now for re-testing all existing tests for classic histograms on nhcb at once.

File: promql/test.go (hunk 1 above)
Comment (krajorama) [reply promql/test.go:739]: Technically we have 5 choices as far as I see:
- explicit tests (easy, but error prone)
- assert that all tests have both versions
- able to mark tests to include in checks with nhcb
- able to mark tests to exclude in checks with nhcb
- do what we do now and fix all cases (not possible)

Do we have access to all `eval` statements? We could avoid having to mark exceptions by just defining the nhcb version explicitly.

Also technically you'd have to remove `le` label from the nhcb expressions as that's the official conversion to native query.

Comment (zenador) [reply promql/test.go:739]: The 2nd one seems a bit complicated. 3rd is slightly more troublesome to implement than the 4th but it is more explicit, I don't really mind 3rd or 4th.

> We could avoid having to mark exceptions by just defining the nhcb version explicitly.
Sorry, what do you mean by this?

> Also technically you'd have to remove le label from the nhcb expressions as that's the official conversion to native query.
Yes it works well even without doing it, but it should be simple enough to just remove it as well.

Comment (zenador) [reply promql/test.go:746]:
```
# More realistic with rates.
eval instant at 50m histogram_quantile(0.2, rate(testhistogram_bucket[5m]))
	{start="positive"} 0.048
	{start="negative"} -0.2

eval instant at 50m histogram_quantile(0.5, rate(testhistogram_bucket[5m]))
	{start="positive"} 0.15
	{start="negative"} -0.15

eval instant at 50m histogram_quantile(0.8, rate(testhistogram_bucket[5m]))
	{start="positive"} 0.72
	{start="negative"} 0.3
```

The above (originally included) tests were initially failing only for the nhcb version, but the bug is now fixed. Note that even without the fix, the following tests were working for both classic and nhcb:

```
# More realistic with rates.
eval instant at 50m histogram_quantile(0.2, rate(testhistogram_bucket{start="positive"}[5m]))
	{start="positive"} 0.048

eval instant at 50m histogram_quantile(0.2, rate(testhistogram_bucket{start="negative"}[5m]))
	{start="negative"} -0.2

eval instant at 50m histogram_quantile(0.5, rate(testhistogram_bucket{start="positive"}[5m]))
	{start="positive"} 0.15

eval instant at 50m histogram_quantile(0.5, rate(testhistogram_bucket{start="negative"}[5m]))
	{start="negative"} -0.15

eval instant at 50m histogram_quantile(0.8, rate(testhistogram_bucket{start="positive"}[5m]))
	{start="positive"} 0.72

eval instant at 50m histogram_quantile(0.8, rate(testhistogram_bucket{start="negative"}[5m]))
	{start="negative"} 0.3
```

Meaning that when you mix the positive and negative series (that have different bucket bounds) together, something gets corrupted somewhere. This isn't a problem for the other tests as the different series have consistent bucket bounds. This isn't a problem for classic histograms or the exponential native histograms, so the problem is with our new customValues field, which we shallow copied when we added it to the model instead of doing a deep copy. I added the fix commit to the earlier chunks PR because I'd rather get the fix in main earlier. We can revisit these optimisations later on when nhcb implementation is more complete and we have more tests.

File: promql/test.go [hunk 3]
```diff
@@ -43,9 +44,16 @@ import (
...
-	patEvalInstant = regexp.MustCompile(`^eval(?:_(fail|ordered))?\s+instant\s+(?:at\s+(.+?))?\s+(.+)$`)
+	patSpace                = regexp.MustCompile("[\t ]+")
+	patLoad                 = regexp.MustCompile(`^load\s+(.+?)$`)
+	patEvalInstant          = regexp.MustCompile(`^eval(?:_(fail|ordered))?\s+instant\s+(?:at\s+(.+?))?\s+(.+)$`)
+	histogramBucketReplacer = strings.NewReplacer(
+		"_bucket", "",
+		" by (le)", "",
+		"(le, ", "(",
+		", le, ", ", ",
+		", le)", ")",
```
Comment (krajorama) [inline promql/test.go:55]: Nice, let's follow the pattern above where whitespace is matched with `\s`

```suggestion
		"\s+by\s*(\s*le\s*)", "",
		"(\s*le\s*,\s*", "(",
		",\s*le\s*,\s*", ", ",
		",\s*le\s*)", ")",
```

File: promql/test.go (hunk 1 above)
Comment (zenador) [reply promql/test.go]: Moved this part to https://github.com/prometheus/prometheus/pull/13827

File: promql/quantile.go [hunk 4]
```diff
@@ -203,14 +203,19 @@ func histogramQuantile(q float64, h *histogram.FloatHistogram) float64 {
 		rank = (1 - q) * h.Count
 	}
 
+	bucketIndex := -1
```
[the remark this answers is not in the document]
Comment (zenador) [reply promql/quantile.go]: was trying to replicate the original logic in bucketQuantile where they check if it's the first/last bucket. added this variable since this one uses an iterator and we can't use the index directly. but you're right that we can probably assume that the bucket with lower -Inf is the first bucket and the bucket with upper +Inf is the last bucket and not bother checking for first/last

File: promql/quantile.go [hunk 5]
```diff
@@ -221,6 +226,17 @@ func histogramQuantile(q float64, h *histogram.FloatHistogram) float64 {
...
 			bucket.Upper = 0
 		}
+	} else if h.UsesCustomBuckets() && bucketIndex == 0 {
+		if bucket.Lower == math.Inf(-1) {
+			// first bucket, with lower bound -Inf
+			if bucket.Upper <= 0 {
+				return bucket.Upper
+			}
+			bucket.Lower = 0
+		} else if bucket.Upper == math.Inf(1) {
```
[the remark this answers is not in the document]
Comment (zenador) [reply promql/quantile.go:234]: this is on the assumption that if we reach the last bucket, q is > 0.5, so we use the reverse iterator and now bucketIndex==0 means the last bucket. but yes we can probably remove this check anyway. i updated `checkHistogramCustomBounds` to check that we don't explicitly include the infinite bound, is that what you meant?

File: promql/quantile.go [hunk 6]
```diff
@@ -221,6 +226,17 @@ func histogramQuantile(q float64, h *histogram.FloatHistogram) float64 {
 			// negative buckets. So we consider 0 to be the upper bound.
 			bucket.Upper = 0
 		}
+	} else if h.UsesCustomBuckets() && bucketIndex == 0 {
+		if bucket.Lower == math.Inf(-1) {
```
[the remark this answers is not in the document]
Comment (zenador) [reply promql/quantile.go:228]: makes sense, removed the check as suggested

File: promql/test.go (hunk 1 above)
Comment (beorn7) [reply promql/test.go]: Thinking more about the test setup, I'm more and more inclined to make everything explicit.

For one, that means that `load`ing a classic histogram needs to have some keyword to activate conversion into NHCB, mostly because there are cases where the test deliberately contains data that cannot be converted to NHCB (see above), _unless_ we decide to tweak the conversion algorithm so that it can deal with those cases (which maybe we should not, see also above). (I think this is an agreement with @zenador's assessment so far.)

However, I also think we should give up on running the classic histogram tests automatically and implicitly with converted NHCB. Here are the reasons I see:
- There are queries that deliberately create a different result, e.g. `rate(histo_bucket)` creates a classic histogram as output, while `rate(histo)` creates a native histogram. There are more subtle cases like the counter reset detection that is deliberately broken in classic histograms with negative observations in the game but works correctly after NHCB conversion. (I don't know if we have a test for that right now, but we could one day or maybe even should.)
- Even for queries that are supposed to create the same output, the conversion isn't trivial. `histogram_quantile(0.5, histo_bucket)` is easy. `histogram_quantile(0.5, sum by (le) (histo_bucket))` is a bit harder. You also have to take into account the also valid `histogram_quantile(0.5, sum(histo_bucket) by (le))` (and there might be more labels in any order and you have to find the `le` label and remove it). Then there are queries involving `foo_count` and `foo_sum`, which needed to be converted to `histogram_count(foo)` and `histogram_sum(foo)`. Etc. In short, I think implementing a correct test converter is a very deep rabbit hole.
- And now add the problem that there are cases where the conversion to NHCB doesn't work by design (see above, where we concluded we need to mark the conversion explicitly anyay).

In summary, I think would need to invest a lot of effort to chase an ultimately impossible goal if we desire to automatically run all classic histogram tests with NHCB, too. In many cases, we need a human anyway to decide if the test should run with NHCB at all, and if yes how the query will look like.

Comment (zenador) [reply promql/test.go]: > maybe even one where the same buckets are defined with different formats (like now) but they have different values. (I'm not even sure what the behavior should be then.

Yeah, it seems like in that case we could arbitrarily select one and the behaviour would be non-deterministic, so perhaps it's better not to test that and just say we provide no guarantees in such edge cases.

> I would currently tend towards (1).

I agree, it is better to only convert valid histograms. If they are problematic, better to flag it out earlier with an error so the user can look into fixing it, than to automatically correct it upon conversion based on assumptions that may be wrong.

Anyway, now both loading and evaluating with native histograms are opt-in, you need to use the keyword suffix `_with_nhcb` for both of those. So we don't have those suffixes for the `nonmonotonic_bucket`, and for `mixed_bucket` we load it, use the keyword for tests with the same results, and don't use the keyword for the failing test, and we now have a separate test for the different result of the nhcb version, so when this disparity gets looked into in future, the tests can be updated to indicate the change in behaviour.

The conversion in the eval query is very simple, and it works well for the current tests that we have, but for more complicated cases like the one you mentioned, it's best that we don't use the keyword and define a separate test instead. I agree that making the test converter completely correct would be more trouble than it's worth, so we'll just have a simple one for convenience to cover most of the simple cases, and fall back to manually defining it when necessary.

File: promql/quantile.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (krajorama) [reply promql/quantile.go:234]: yes. That's how we defined it in remote write as well:
```
// The last element is not only the upper inclusive bound of the last regular
  // bucket, but implicitly the lower exclusive bound of the +Inf bucket.
```

File: promql/test.go [hunk 7]
```diff
@@ -416,6 +432,157 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+		Set(labels.MetricName, strings.TrimSuffix(mName, suffix)).
+		Del(labels.BucketLabel).
+		Labels()
+	hash := baseM.Hash()
+	return baseM, hash
+}
+
+// if classic histograms are defined, convert them into native histograms with custom
+// bounds and append the defined time series to the storage.
+func (cmd *loadCmd) appendCustomHistogram(a storage.Appender) error {
```
Comment (krajorama) [inline promql/test.go:453]: This algorithm is pretty complicated.
I'm thinking that this could be simpler if we kept a map[uint64]map[float]*histogram.Histogram. 
Where we map from hash of series (without suffix or `le` label)  to   map of timestamp to the resulting histogram.


So when you hit a _count or _sum series, you remove the suffix, take the hash and find/create the histogram at the timestamp and set count or sum.
When you hit a _bucket, you'd again remove the suffix and `le` label and find/create the histogram at the timestamp and *insert* the bound as appropriate at `le` value into CustomValues (incl. +Inf for now)  and also the bucket count into PositiveBuckets as an absolute value at the same position as CustomValues.

When all is done, you take another pass and turn the absolute values into deltas in PositiveBuckets and remove +Inf from the custom bounds. And also check the overall count and do any other check.

File: promql/test.go [hunk 8]
```diff
@@ -43,10 +44,17 @@ import (
...
-	patLoad        = regexp.MustCompile(`^load\s+(.+?)$`)
-	patEvalInstant = regexp.MustCompile(`^eval(?:_(fail|ordered))?\s+instant\s+(?:at\s+(.+?))?\s+(.+)$`)
-	patEvalRange   = regexp.MustCompile(`^eval(?:_(fail))?\s+range\s+from\s+(.+)\s+to\s+(.+)\s+step\s+(.+?)\s+(.+)$`)
+	patSpace                = regexp.MustCompile("[\t ]+")
+	patLoad                 = regexp.MustCompile(`^load(?:_(with_nhcb))?\s+(.+?)$`)
+	patEvalInstant          = regexp.MustCompile(`^eval(?:_(with_nhcb))?(?:_(fail|ordered))?\s+instant\s+(?:at\s+(.+?))?\s+(.+)$`)
+	patEvalRange            = regexp.MustCompile(`^eval(?:_(fail))?\s+range\s+from\s+(.+)\s+to\s+(.+)\s+step\s+(.+?)\s+(.+)$`)
+	histogramBucketReplacer = strings.NewReplacer(
+		"_bucket", "",
+		" by (le)", "",
```
Comment (krajorama) [inline promql/test.go:53]: Let's use `\s+` or `\s*` for whitespace as appropriate, see examples above.

Comment (zenador) [reply promql/test.go:53]: It's a bit different as strings pkg doesn't do regex AFAIK, but I'll try to find a regex equivalent.

File: promql/test.go (hunk 7 above)
Comment (zenador) [reply promql/test.go:453]: I see what you mean about it being messy right now, but I feel that continually inserting bounds and counts and updating the spans in the histogram is also kind of messy, and while it's going through this process the histogram would be very invalid, i.e. total count and bucket counts will be wrong since you need to make it uncumulative afterwards, and you'd have an +Inf bound that isn't supposed to be there. How about I create a temporary struct just for this instead, then the second pass can convert that temp struct into a histogram and check validity?

Comment (krajorama) [reply promql/test.go:453]: I think a temporary struct is good compromise.

File: promql/test.go [hunk 9]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+
+	// Go through all the time series to collate classic histogram data
+	// and organise them by timestamp.
+	for hash, smpls := range cmd.defs {
+		m := cmd.metrics[hash]
+		mName := m.Get(labels.MetricName)
+		switch {
+		case strings.HasSuffix(mName, "_bucket") && m.Has(labels.BucketLabel):
+			le, err := strconv.ParseFloat(m.Get(labels.BucketLabel), 64)
+			if err != nil {
```
Comment (krajorama) [inline promql/test.go:531]: Nothing is equal to NaN and also NaN is not a valid boundary, so we can use later.

```suggestion
			if err != nil || math.IsNaN(le) {
```

File: promql/test.go [hunk 10]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+			})
+		}
+	}
+
+	// Convert the collated classic histogram data into native histograms
+	// with custom bounds and append them to the storage.
+	for _, histWrapper := range histMap {
+		sort.Float64s(histWrapper.upperBounds)
+		upperBounds := make([]float64, 0, len(histWrapper.upperBounds))
+		prevLe := -372.4869 // assuming that the lowest bound is never this value
```
Comment (krajorama) [inline promql/test.go:555]:
```suggestion
		prevLe := math.NaN()
```

File: promql/test.go [hunk 11]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+		}
+		fhBase := &histogram.FloatHistogram{
+			Count:  0,
+			Sum:    0,
+			Schema: histogram.CustomBucketsSchema,
+			PositiveSpans: []histogram.Span{
+				{Offset: 0, Length: uint32(len(upperBounds))},
+			},
+			CustomValues: customBounds,
+		}
```
Comment (krajorama) [inline promql/test.go:576]: line 553-576 is ripe for a helper function that just creates the base histogram with the custome bounds set

File: promql/test.go [hunk 12]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+		ts := make([]int64, 0, len(histWrapper.histByTs))
+		for t := range histWrapper.histByTs {
+			ts = append(ts, t)
+		}
+		sort.Slice(ts, func(i, j int) bool { return ts[i] < ts[j] })
```
Comment (krajorama) [inline promql/test.go:581]: I think it would be simpler to just create the slice of `Sample`s  unsorted , call sort.Slice on that, before calling appendSample. So this pre-sorting into `ts` can be removed altogether.

File: promql/test.go [hunk 13]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+		for _, t := range ts {
```
Comment (krajorama) [inline promql/test.go:582]: As mentioned above, let's just iterate through `histWrapper.histByTs`.

File: promql/test.go [hunk 14]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+			hist, exists := histWrapper.histByTs[t]
+			if !exists {
+				continue
+			}
```
Comment (krajorama) [inline promql/test.go:586]: This will be gone.

File: promql/test.go [hunk 15]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+			customBounds = upperBounds
+		}
+		fhBase := &histogram.FloatHistogram{
+			Count:  0,
+			Sum:    0,
+			Schema: histogram.CustomBucketsSchema,
+			PositiveSpans: []histogram.Span{
+				{Offset: 0, Length: uint32(len(upperBounds))},
+			},
+			CustomValues: customBounds,
```
Comment (krajorama) [inline promql/test.go:575]: You can also allocate `PositiveCounts` and fill with zeros.

All functions must work with buckets that have zero counts.
I don't know how we'll do scrape, we'll have to revisit this to when that's done.
I think the safest is probably to call fh.Compact() after bucket counts are updated from zeros. Whatever we'll do we should reflect here in this function to follow scrape.

File: promql/test.go [hunk 16]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+			}
+			fh.Sum = hist.sum
+			if hist.count != 0 {
+				total = hist.count
+			}
+			fh.Count = total
+			fh.PositiveBuckets = counts
+			// s := Sample{Metric: m, T: t, H: fh}
+			// fmt.Printf("end s: %v\n", s)
+			s := Sample{T: t, H: fh}
```
Comment (krajorama) [inline promql/test.go:608]: Should probably call Compact() as discussed above.

```suggestion
			s := Sample{T: t, H: fh.Compact()}
```

File: promql/test.go [hunk 17]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+			ts = append(ts, t)
+		}
+		sort.Slice(ts, func(i, j int) bool { return ts[i] < ts[j] })
+		for _, t := range ts {
+			hist, exists := histWrapper.histByTs[t]
+			if !exists {
+				continue
+			}
+			fh := fhBase.Copy()
+			counts := make([]float64, 0, len(upperBounds))
```
Comment (krajorama) [inline promql/test.go:588]:
```suggestion
```

File: promql/test.go [hunk 18]
```diff
@@ -416,6 +450,171 @@ func (cmd *loadCmd) append(a storage.Appender) error {
...
+			var prevCount, total float64
+			for _, le := range upperBounds {
+				currCount, exists := hist.bucketCounts[le]
+				if !exists {
+					currCount = 0
+				}
+				count := currCount - prevCount
```
Comment (krajorama) [inline promql/test.go:595]: FloatHistogram has absolute counts, so I'm confused here. You just need to find the index where to write the absolute value.

Comment (krajorama) [reply promql/test.go:595]: never mind, forgot what we convert from

File: promql/test.go (hunk 10 above)
Comment (zenador) [reply promql/test.go]: I ended up making this -Inf instead, since that should also not be a valid boundary.

**Review by krajorama — APPROVED** [review]
